doc: note about non-existent files in watchFile

thefourtheye opened the pull request:
Refer: https://github.com/nodejs/io.js/pull/2093#discussion_r34343965

cc @brendanashworth @bnoordhuis

thefourtheye commented:
Bump!

silverwind commented on doc/api/fs.markdown:
```diff
@@ -577,7 +577,12 @@ If you want to be notified when the file was modified, not just accessed
+ If the file got created later on, the listener will be called again, with the
+ latest stat objects._
+
+_This is a change in functionality since v0.10._
```
Move that up to be included in above paragraph?

brendanashworth commented on doc/api/fs.markdown:
```diff
@@ -577,7 +577,12 @@ If you want to be notified when the file was modified, not just accessed
 _Note: when an `fs.watchFile` operation results in an `ENOENT` error, it will
-invoke the callback once. This is a change in functionality since v0.10._
+ invoke the listener once, with two stat objects, with all the fields set to
+ zero, except for the time fields, which are set to Thu Jan 01 1970 00:00:00.
```
This is a run on sentence, that is a little bit confusing, with all the clauses. :)

thefourtheye commented:
@brendanashworth @silverwind I incorporated the review comments. PTAL now :-)

brendanashworth commented on doc/api/fs.markdown:
```diff
@@ -577,7 +577,11 @@ If you want to be notified when the file was modified, not just accessed
 
 _Note: when an `fs.watchFile` operation results in an `ENOENT` error, it will
-invoke the callback once. This is a change in functionality since v0.10._
+ invoke the listener once, with two stat objects. All the fields in those stat
```
Since `two stat objects` is redundant - it always does two stat objects - you could change it to something like `invoke the listener once, with all fields zeroed (or, for dates, the Unix Epoch).`

brendanashworth commented on doc/api/fs.markdown:
```diff
@@ -577,7 +577,11 @@ If you want to be notified when the file was modified, not just accessed
-invoke the callback once. This is a change in functionality since v0.10._
+ invoke the listener once, with two stat objects. All the fields in those stat
+ objects will be set to zero, except for the time fields, which are set to
+ Thu Jan 01 1970 00:00:00. If the file got created later on, the listener will
```
`s/got created/was created`? "got" is more colloquial and I'd say its better to use the alternative in docs.

thefourtheye commented:
@brendanashworth Thanks :-) Updated. PTAL now.

brendanashworth commented:
LGTM. Did you want to add a test for this [here](https://github.com/nodejs/io.js/blob/ab479659c789f938f0560b49167ba85ef825f1ba/test/parallel/test-fs-watchfile.js#L24)?

thefourtheye commented:
@brendanashworth Oh yeah, Thanks for pointing out :-) I included a test case now. PTAL.

thefourtheye commented:
CI Run: [iojs+any-pr+multi/217](https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/217/)

thefourtheye commented:
@bnoordhuis Looks like, in Windows `blocks` and `blksize` fields are `undefined` instead of zero. Shall I specialize those fields for Windows in test and doc?

bnoordhuis commented:
Oh, that sounds right.  Yes, special-casing for Windows is okay.

thefourtheye commented:
New CI run with Windows specific changes: [iojs+any-pr+multi/219](https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/219/)

brendanashworth commented:
LGTM if the CI is happy

thefourtheye commented:
New CI run: [node-test-commit/85](https://jenkins-iojs.nodesource.com/job/node-test-commit/85/)

thefourtheye commented:
The build is Green. @bnoordhuis LGTY?

bnoordhuis commented on test/parallel/test-fs-watchfile.js:
```diff
@@ -19,8 +19,74 @@ assert.throws(function() {
+
+function areEmptyAndEqual(curr, prev) {
+  // create copies of arguments without the prototype information, because
+  // assert.deepStrictEqual compares the prototypes as well.
```
That shouldn't matter because the prototypes are the same, I think?

thefourtheye replied:
@bnoordhuis Actually we are comparing `expected` against `curr` and `prev`. `expected` is just an object literal. So, the prototypes will be different.

bnoordhuis replied:
Can't you do `const expected = new fs.Stats()`?

thefourtheye replied:
@bnoordhuis It will use `undefined` for all the values.

    > new fs.Stats()
    { dev: undefined,
      mode: undefined,
      nlink: undefined,
      uid: undefined,
      gid: undefined,
      rdev: undefined,
      blksize: undefined,
      ino: undefined,
      size: undefined,
      blocks: undefined,
      atime: Invalid Date,
      mtime: Invalid Date,
      ctime: Invalid Date,
      birthtime: Invalid Date }

Otherwise, I should be doing something like this

    > const sDate = Date.UTC(1970, 0, 1, 0, 0, 0);
    undefined
    > new fs.Stats(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, sDate, sDate, sDate, sDate);
    { dev: 0,
      mode: 0,
      nlink: 0,
      uid: 0,
      gid: 0,
      rdev: 0,
      blksize: 0,
      ino: 0,
      size: 0,
      blocks: 0,
      atime: Thu Jan 01 1970 05:30:00 GMT+0530 (IST),
      mtime: Thu Jan 01 1970 05:30:00 GMT+0530 (IST),
      ctime: Thu Jan 01 1970 05:30:00 GMT+0530 (IST),
      birthtime: Thu Jan 01 1970 05:30:00 GMT+0530 (IST) }

bnoordhuis replied:
I took a shortcut there, sorry, but the second one is indeed what I had in mind.

thefourtheye replied:
@bnoordhuis Cool, I updated the test like that.

bnoordhuis commented on test/parallel/test-fs-watchfile.js:
```diff
@@ -19,8 +19,74 @@ assert.throws(function() {
+
+// If the file initially didn't exist, and gets created at a later point of
+// time, the callback should be invoked again with proper values in stat object
+var fileExists = false;
```
I'd replace `fileExists` with a counter and check that it never >= 2.

thefourtheye commented:
@bnoordhuis I updated the test to use `fs.Stats` to create the expected object. PTAL.

@brendanashworth I rebased and it is a single commit now :-) I feel that it would be better if you reviewed it once again.

CI Run: [node-test-pull-request/42/](https://jenkins-iojs.nodesource.com/job/node-test-pull-request/42/)

thefourtheye commented:
Build is green, except the linter. I ran `make lint` and it didn't complain.

jbergstroem commented:
Linter back now. Run here: https://jenkins-iojs.nodesource.com/job/node-test-pull-request/43/

thefourtheye commented:
Thanks @jbergstroem :-) Linter also has given Green signal :-)

bnoordhuis commented:
LGTM with a nit about the commit log: the first line should be <= 50 characters.

Aside, I would reword this:

> this patch improves the Note section in watchFile to clarify about the non-existent files.

To e.g. this:

> this patch documents the behavior of calling fs.watchFile() with a path that does not yet exist.

thefourtheye commented:
@bnoordhuis I am landing it now and I ll change it

thefourtheye commented:
Thanks guys, landed at ed85c95a9c252071e91bccfaf854d23a2ebe9a2b.
